logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb

```diff
@@ -37,5 +37,231 @@ def take
     def poll(millis)
       @queue.poll(millis, TimeUnit::MILLISECONDS)
     end
+
+    def write_client
+      WriteClient.new(self)
+    end
+
+    def read_client
+      ReadClient.new(self)
+    end
+
+    class ReadClient
+      # We generally only want one thread at a time able to access pop/take/poll operations
+      # from this queue. We also depend on this to be able to block consumers while we snapshot
+      # in-flight buffers
+
+      def initialize(queue)
+        @queue = queue
+        @mutex = Mutex.new
+        @inflight_batches = {}
+        @batch_size = 125
+        @wait_for = 5
+      end
+
+      def set_batch_details(batch_size, wait_for)
+        @batch_size = batch_size
+        @wait_for = wait_for
+      end
+
+      def set_events_metric(metric)
+        @event_metric = metric
+      end
+
+      def set_pipeline_metric(metric)
+        @pipeline_metric = metric
+      end
+
+      def inflight_batches
+        @mutex.synchronize do
+          yield(@inflight_batches)
+        end
+      end
+
+      def current_inflight_batch
+        @inflight_batches.fetch(Thread.current, [])
+      end
```

Comment:
I added that in for future debugging purposes. I think its safe to remove it now if we aren't using it.